File: day_3/day_3_solution_2.rb
def test_path(row_inc, col_inc, lines)
    num_cols = lines[0].length
    num_rows = lines.length()
    
    cur_col = 0
    cur_row = 0
    tree_count = 0

    while (cur_row < num_rows)
        adjusted_col = cur_col % num_cols
        char = lines[cur_row][adjusted_col]

#        print "Row: #{cur_row}, Col: #{adjusted_col}, Char: #{char}\n";

        if (char == '#')
            tree_count += 1
        end

        cur_row += row_inc
        cur_col += col_inc
    end
    print "Tree Count: #{tree_count}\n"
    return tree_count
end

lines = File.read("input_3.txt").split
paths = [[1, 1], [3, 1], [5, 1], [7, 1], [1, 2]];

result = 1
for path in paths
    result *= test_path(path[1], path[0], lines)
end
print "Result: #{result}\n"

File: 2020/day_3/day_3_solution_2.pl
sub test_path{
    my (@ops) = @_;
    my $row_increment = $ops[0];
    my $col_increment = $ops[1];
    my @lines = @{$_[2]};
    
    $num_cols = length $lines[0];
    $num_rows = $#lines;
    
    $cur_col = 0;
    $cur_row = 0;

    my $tree_count = 0;

    while ($cur_row <= $num_rows){
        $adjusted_col = $cur_col % $num_cols;
        $char = substr($lines[$cur_row], $adjusted_col, 1);

        #print "Row: $cur_row, Col: $adjusted_col, Char: $char\n";

        if ($char eq '#'){
            $tree_count += 1;
        }

        $cur_row += $row_increment;
        $cur_col += $col_increment;
    }
    print "Tree Count: $tree_count\n";
    return $tree_count;
}

open(inputfile, "input_3.txt");
chomp(my @lines = <inputfile>);

@paths = ([1, 1], [3, 1], [5, 1], [7, 1], [1, 2]);

for $path (0 .. $#paths){
    $tree_counts[$path] = test_path($paths[$path][1], $paths[$path][0], \@lines);
}

$answer = $tree_counts[0] * $tree_counts[1] * $tree_counts[2] * $tree_counts[3] * $tree_counts[4];
print "Answer: $answer\n"